### ISSUE TITLE
building and running

### ISSUE THREAD

**Issue by drdozer** [stgatilov/position-counting-sort#1]
Hi -- I'm currently looking at optimised sorting and found your blog and this source. Can you give me a hint as to how to build/run it please? I'm on ubuntu, if that helps.

**Comment by stgatilov** [comment]
Added [shell script](https://github.com/stgatilov/position-counting-sort/blob/master/c.sh).
Basically, you simply throw the only cpp-file into g++, and allow using SSE4. Don't forget optimization, or it will be slow -- and that's it =)

Note that various sizes/options are specified at the beginning of "testing" section in cpp file.

**Comment by drdozer** [comment]
Thanks. that's worked. c++ isn't really my thing. I'm learning rust at the moment and trying to get a high performance, large data sort working. Your code is really helpful.

**Comment by drdozer** [comment]
So just so that I understand, it gave this output:

```
umber of elements = 32
   166.6 ns : PCSort_Main:any                            (104505344)
   104.8 ns : PCSort_Optimized:any                       (104505344)
   100.5 ns : PCSort_Trans:any                           (104505344)
   103.9 ns : PCSort_Main:distinct                       (1266135040)
    73.6 ns : PCSort_WideOuter:distinct                  (1266135040)
    75.7 ns : SortingNetwork_32                          (104902656)
   315.6 ns : InsertionSort                              (104902656)
   421.1 ns : StdSort                                    (104902656)
```

Does that mean that on my particular hardware (Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz) that PCSort_WideOuter:distinct and SortingNetwork_32 are performing the best?

I've implemented a 32 element sorting network in rust, and it's coming in at approximately half that performance, with a median execution time of 126 ns, so they must be producing subtly different assembly.

**Comment by stgatilov** [comment]
Yes, it seems that on your machine sorting network is faster than on mine. You should ensure that cmovs are generated instead of branches.

Anyway, it won't help you much with large-array sorting. This code is tuned for small sizes only. For big arrays, you should search for radix sort: both LSB and MSB version (which is also called bucket sort). Cache performance and number of passes are more important.

**Comment by drdozer** [comment]
I was thinking of using an adapted pivot-sort to make blocks of 32 elements that could then be sorted in-place as a block. Thinking of the distribution of my data, I'm not sure that a radix sort would be ideal. The buckets would be very unevenly sized.

**Comment by stgatilov** [comment]
You can use traditional LSD version of radix sort. It does not care about distribution, and is much faster than quicksort.
